configuration/postgres/postgres.go

```diff
@@ -0,0 +1,401 @@
+)
+
+func NewPostgresConfigurationStore(logger logger.Logger) configuration.Store {
+	logger.Debug("Instantiating PostgreSQL configuration store")
+	return &ConfigurationStore{
+		logger: logger,
+	}
+}
+
+func (p *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	p.logger.Debug(InfoStartInit)
+
+	if p.client != nil {
+		return fmt.Errorf(ErrorAlreadyInitialized)
+	}
+	if m, err := parseMetadata(metadata); err != nil {
+		p.logger.Error(err)
+		return err
+	} else {
+		p.metadata = m
+	}
+	p.ActiveSubscriptions = make(map[string]*subscription)
+	ctx, cancel := context.WithTimeout(context.Background(), p.metadata.maxIdleTime)
+	defer cancel()
+	client, err := Connect(ctx, p.metadata.connectionString, p.metadata.maxIdleTime)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	pingErr := p.client.Ping(ctx)
+	if pingErr != nil {
+		return pingErr
+	}
+	// check if table exists
+	exists := false
+	err = p.client.QueryRow(ctx, QueryTableExists, p.metadata.configTable).Scan(&exists)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequest) (*configuration.GetResponse, error) {
+	if err := validateInput(req.Keys); err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	query, params, err := buildQuery(req, p.metadata.configTable)
+	if err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	rows, err := p.client.Query(ctx, query, params...)
+	if err != nil {
+		// If no rows exist, return an empty response, otherwise return the error.
+		if err == sql.ErrNoRows {
+			return &configuration.GetResponse{}, nil
+		}
+		return nil, err
+	}
+	items := make(map[string]*configuration.Item)
+	for i := 0; rows.Next(); i++ {
+		var item configuration.Item
+		var key string
+		var metadata []byte
+		v := make(map[string]string)
+
+		if err := rows.Scan(&key, &item.Value, &item.Version, &metadata); err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(metadata, &v); err != nil {
+			return nil, err
+		}
+		item.Metadata = v
+		if item.Value != "" {
+			items[key] = &item
+		}
+	}
+	return &configuration.GetResponse{
+		Items: items,
+	}, nil
+}
+
+func (p *ConfigurationStore) Subscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler) (string, error) {
+	var triggers []string
+	for k, v := range req.Metadata {
+		if res := strings.EqualFold("trigger", k); res {
+			triggers = append(triggers, v)
+		}
+	}
+
+	if len(triggers) == 0 {
+		return "", fmt.Errorf("cannot subscribe to empty trigger")
+	}
+
+	var subscribeID string
+	for _, trigger := range triggers {
+		key := "listen " + trigger
+		if sub, isActive := p.isSubscriptionActive(req); isActive {
+			if oldStopChan, ok := p.subscribeStopChanMap.Load(sub); ok {
+				close(oldStopChan.(chan struct{}))
+			}
+		}
+		stop := make(chan struct{})
+		subscribeID = uuid.New().String()
+		p.subscribeStopChanMap.Store(subscribeID, stop)
```

Comment:
I see that here you're using a sync.Map. However, because there are other things in this method that are not safe for concurrent use, I'd recommend using a regular map here, and wrapping the entire Subscribe/Unsubscribe methods in a lock.

Reply:
https://github.com/dapr/components-contrib/blob/4f64eefae64507ca5cfd5f11044283816b126d4a/configuration/postgres/postgres.go#L188

Reply:
done